sdk/core/azure-core/src/http/curl/curl.cpp

```diff
@@ -180,6 +180,56 @@ static inline std::string GetHTTPMessagePreBody(Azure::Core::Http::Request const
 
   return httpRequest;
 }
+
+static void CleanupThread()
+{
+  using namespace Azure::Core::Http::_detail;
+  for (;;)
+  {
+
+    // Won't continue until the ConnectionPoolMutex is released from MoveConnectionBackToPool
+    std::unique_lock<std::mutex> lockForPoolCleaning(
+        CurlConnectionPool::g_curlConnectionPool.ConnectionPoolMutex);
+
+    // Wait the defined default time OR to the signal from the conditional variable.
+    // wait_for releases the mutex lock until it wakes up again or it's cancelled.
+    if (CurlConnectionPool::g_curlConnectionPool.ConditionalVariableForCleanThread.wait_for(
+            lockForPoolCleaning,
+            std::chrono::milliseconds(DefaultCleanerIntervalMilliseconds),
+            []() {
+              return CurlConnectionPool::g_curlConnectionPool.ConnectionPoolIndex.size() == 0;
+            }))
+    {
+      // Cancelled by another thead or no connections on wakeup
+      return;
+    }
+
+    // loop the connection pool index - Note: lock is re-taken for the mutex
+    // Notes:
+    // - If all connections are removed from a host-index, the index is not removed from the pool to
+    // reduce the time of the entire routine.
+    // - A host-index with size 0 (no connections) is ignored.
+    for (auto index = CurlConnectionPool::g_curlConnectionPool.ConnectionPoolIndex.begin();
+         index != CurlConnectionPool::g_curlConnectionPool.ConnectionPoolIndex.end()
+         && index->second.size() > 0;
+         index++)
+    {
+      // Pool index with waiting connections. Each pool index behaves as a Last-in-First-out list,
+      // the last connection moved to the pool will be the first to be re-used. Because of this, the
+      // oldest connection in the pool can be found at the end of the list. Looping the connection
+      // pool backwards until a connection that is not expired is found or until all connections are
+      // removed.
+      for (auto connection = --(index->second.end());
+           index->second.size() > 0 && connection->get()->IsExpired();
```

Comment:
Re-word and keep the original comments you had earlier to explain the for-loop signature.
`// Connection removed, break if there are no more connections to check`

`// Got a non-expired connection, all connections before this one are not`
`// expired. Break the loop and continue looping the Pool index`